src/librustc_codegen_llvm/debuginfo/metadata.rs

```diff
@@ -1361,11 +1361,11 @@ impl EnumMemberDescriptionFactory<'ll, 'tcx> {
                         let niche_value = if i == dataful_variant {
                             None
                         } else {
-                            let niche = (i as u128)
+                            let value = (i as u128)
                                 .wrapping_sub(*niche_variants.start() as u128)
                                 .wrapping_add(niche_start);
-                            assert_eq!(niche as u64 as u128, niche);
-                            Some(niche as u64)
+                            let value = value & ((1u128 << niche.value.size(cx).bits()) - 1);
```

Comment:
Removing the assert is wrong, because you're throwing away bits just below!

WIth the new change, `assert_eq!(niche as u64 as u128, niche);` should be between the truncation (`let value = value & ...;`) and the lossy cast (`value as u64`).